### PR TITLE
Unit 2 Linear Algebra

### PR DESCRIPTION
**Pull request by DanielleRomanoff**
My pull request for this Unit

### CODE REVIEW

**Comment by jojordan3** [comment]
## Sprint Challenge Code Review
### Unit 2: Week 1 - Linear Algebra

---
#### Part 1 — Linear Algebra
<details><summary><em>Code review</em></summary>
<p>

_What went well:_

- This looks like it was a walk through the park for you. Textbook solutions to the questions.

_What could be better:_

- Wasn't really necessary here because it was only asking for 1 thing in each cell, but it's good to get into the habit of including labels within your print statements. You do it in section 2, but it'd be to your benefit if that just became automatic whenever you print out a value.

</p>
</details>

---
#### Part 2 — Intermediate Linear Algebra
<details><summary><em>Code review</em></summary>
<p>

_What went well:_

- You calculated everything correctly here, and you have a good explanation for 2.2

_What could be better:_

- You did a lot of extra work in this section. Really try to use built in functions. Both Pandas and NumPy have `.std()`, `.mean()`, `.var()`, and `.cov()` functions. We're going to have to get you out of the habit of trying to do all the math yourself :). It's not just because it takes more lines to do it, but when you get to the point where you're running long programs, computational efficiency matters, and functions in these well-established libraries are generally highly optimized. I know it's a mind shift and it'll be a process, but I'll be here to remind you. 😄 

</p>
</details>

---
#### Part 3 — PCA
<details><summary><em>Code review</em></summary>
<p>

_What went well:_

- Your conclusion/explanation is spot on, and it's great that you are thinking about what this all means as you do it (not everyone is 😉 )

- You have all the right pieces here. You completed all the separate steps... 

_What could be better:_

- The only problem is you completed them separately, a little out of order, and you did PCA twice as well as standardized the dataset twice. This may be where you second-guessed yourself because it does look like this was redone kind of... cyclically. 

- Here's a good example of why you should use built-in functions whenever you can:
```python
normalized =(X - X.mean()/X.std())
```
This did not get you the correct result... And I assume at some point you had to hound your geo or calc students about making sure they put alllll the parentheses they needed into their graphing calculator 🙂 

- You used correct function to SciKit function to standardize the data, but you should use it straight on the raw data:
```python
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()

scaler.fit(X)
food = scaler.transform(X)
```
Then you should put `food` straight into the PCA
```python
from sklearn.decomposition import PCA
# create the PCA instance
pca = PCA(n_components=2)
# fit on data
pca.fit(food)
X_pca = pca.transform(food)
```
Then it wasn't necessary here, but it is good practice to look at the eigenvectors and eigenvalues like you did.
Notice, that now, `X_pca` *is* the matrix that has the points you want to graph. You can stick that right into the `pd.DataFrame()` function. So you could put:
```python
# Using the "columns" parameter is a nice way to skip making the dictionary and go straight to the
# DataFrame
X_df = pd.DataFrame(X_pca, columns=["x", "y"]) 
plt.scatter(X_df['x'], X_df['y']) # This is all you needed to graph
~~plt.scatter(X_pca['x'], X_pca['y'])~~
plt.title("Data After PCA")
plt.xlabel('PC1')
plt.ylabel('PC2')
plt.show()
```
You seem to go back to earlier, unprocessed data as you move on to another portion, so just keep in mind that the point of the standardization is to optimize PCA.

</p>
</details>

---
#### Part 4 — Clustering
<details><summary><em>Code review</em></summary>
<p>

_What went well:_

- I love that you did do the elbow method and explored the data by visualizing it in a graph to get an idea for what k should be before jumping into clustering. You code in these tasks is very efficient. 

- Again you have all the right pieces...

_What could be better:_

- I definitely see you did some going back and changing things because the variable you put into your K Means function does not appear anywhere else in the notebook anymore. It was still defined because you didn't restart your runtime, so you didn't get an error, but if you tried after opening it up again it would break. PCA was not needed for clustering here, especially since you only had 2 dimensions to begin with. Just replace
```python
kmeans.fit(points_pca)
```
with
```python
kmeans.fit(points)
```
That's why your centroids are off. You plotted the clusters themselves correctly. 

</p>
</details>

---
#### Additional notes
<details><summary><em>Code review</em></summary>
<p>

Also, just a suggestion to save you time... and maybe keep you from breaking your code by restarting runtime. 
Skim through the assignment beforehand to see what libraries you might need to update and do that all in a cell at the very beginning of the notebook.

</p>
</details>
